pkg/operator/target_status.go

```diff
@@ -0,0 +1,973 @@
+
+	patchPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.PodMonitoring) error
+	patchClusterPodMonitoring(ctx context.Context, podMonitoring *monitoringv1.ClusterPodMonitoring) error
+
+	// Returns the target fetcher, with an zero-index number `thread` that signifies which goroutine this is for.
+	getTargetFetcher(ctx context.Context, thread uint16) (targetFetcher, error)
+}
+
+// Responsible for fetching the targets given a pod. One is created per goroutine.
+type targetFetcher interface {
+	getTarget(ctx context.Context, logger logr.Logger, port int32, pod *corev1.Pod) (*prometheusv1.TargetsResult, error)
+}
+
+type targetStatusProcessor struct {
+	logger             logr.Logger
+	opts               Options
+	dataManagerBuilder dataManagerBuilder
+}
+
+func NewTargetStatusPoller(logger logr.Logger, restConfig *rest.Config, registry prometheus.Registerer, opts Options) (TargetStatusProcessor, error) {
+	builder, err := newInformerDataManagerBuilder(restConfig, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return newTargetStatusPoller(logger, opts, registry, builder)
+}
+
+func newTargetStatusPoller(logger logr.Logger, opts Options, registry prometheus.Registerer, builder dataManagerBuilder) (TargetStatusProcessor, error) {
+	if err := registry.Register(targetStatusDuration); err != nil {
+		return nil, err
+	}
+
+	instance := &targetStatusProcessor{
+		logger:             logger,
+		opts:               opts,
+		dataManagerBuilder: builder,
+	}
+	return instance, nil
+}
+
+func (p *targetStatusProcessor) Start(ctx context.Context, tickerBuilder func() Ticker) error {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	manager, err := p.dataManagerBuilder.build(stopCh)
+	if err != nil {
+		return err
+	}
+
+	ticker := tickerBuilder()
+	defer ticker.Stop()
+	mutex := newTryMutex()
+
+	onceLog := new(sync.Once)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-ticker.Channel():
+			// Wrapper function so we could call defer within it.
+			err := func() error {
+				// Don't allow multiple polling at the same time.
+				if !mutex.TryLock() {
+					p.logger.Info("Target status poll in progress, skipping")
+					return nil
+				}
+				defer mutex.Unlock()
+
+				now := time.Now()
+				if err := poll(ctx, p.logger, p.opts, manager); err != nil {
+					// Not having data is not an error.
+					if apierrors.IsNotFound(err) {
+						onceLog.Do(func() {
+							p.logger.Error(err, "Kubernetes config not found")
+						})
+						return nil
+					}
+					return err
+				}
+				// In case Kubernetes config is removed and re-added.
+				onceLog = new(sync.Once)
+
+				duration := time.Since(now)
+				targetStatusDuration.WithLabelValues().Set(float64(duration.Milliseconds()))
+				return nil
+			}()
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func poll(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) error {
+	targets, err := fetchTargets(ctx, logger, opts, manager)
```

Comment:
same question on the private func, are `fetchTargets` and `processTargets` used or tested anywhere else or could they just be inlined?

Reply:
`fetchTargets` is unit tested. I will inline the latter.